app/services/samples/metadata/update_service.rb

```diff
@@ -58,44 +61,48 @@ def transform_metadata_keys
       end
 
       def perform_metadata_update
-        update_status = { updated: [], not_updated: [] }
         @metadata.each do |key, value|
           if value.blank?
             if @sample.metadata.key?(key)
               @sample.metadata.delete(key)
               @sample.metadata_provenance.delete(key)
-              update_status[:updated].append(key)
+              @metadata_changes[:deleted] << key
             end
           else
-            metadata_updated = assign_metadata_to_sample(key, value)
-            metadata_updated ? update_status[:updated].append(key) : update_status[:not_updated].append(key)
+            assign_metadata_to_sample(key, value)
           end
         end
-        update_status
+        @update_status
       end
 
-      def assign_metadata_to_sample(key, value)
+      def assign_metadata_to_sample(key, value) # rubocop:disable Metrics/AbcSize
         # We don't overwrite existing @sample.metadata_provenance or @sample.metadata
         # that has a {source: 'analysis'} with a user
         if @sample.metadata_provenance.key?(key) && @analysis_id.nil? &&
            @sample.metadata_provenance[key]['source'] == 'analysis'
-          false
+          @metadata_changes[:not_updated] << key
         else
+          @sample.metadata.key?(key) ? @metadata_changes[:updated] << key : @metadata_changes[:added] << key
           @sample.metadata_provenance[key] =
             @analysis_id.nil? ? { source: 'user', id: current_user.id } : { source: 'analysis', id: @analysis_id }
           @sample.metadata[key] = value
-          true
         end
       end
 
-      def fields_not_updated(metadata_fields_update_status)
-        metadata_fields_not_updated = metadata_fields_update_status[:not_updated]
+      def assign_errors_to_not_updated_fields
```

Comment:
Since this only gets called if there are `not_updated` keys found in the `assign_metadata_to_sample` function, I think this function should be renamed. It suggests that errors raised by the other validation functions (like sample not on project) would get added to the not updated list.

Something like `handle_not_updated_fields` would be good. And probably a good idea to add a function comment saying it only handles the not updated keys found by `assign_metadata_to_sample` function.

Reply:
I like this a lot better. Changed in 422c2be245bdf39d1f2475382065ee11a0c855e5